### DISCUSSION TITLE
Code Generation

### DISCUSSION THREAD

**Discussion by Kamal-a-1** [casadi/casadi#3302] [Q&A]
Hi all,

I am trying to generate a C code from a MATLAB code (MPC) that uses CasADi. I initially tried to use MATLAB Coder to do it, but it couldn't generate code from casadi.Function() because it contains some statements that are not supported by MATLAB Coder.

Therefore I am trying to use the code generation feature of Casadi itself, but from the user guide and the examples, what I understand is that it only generates C code from a function, is it correct?
That's why I am confused about how to generate a C code. The rough structure of the MATLAB code is the following (only the parts that use Casadi):

```
ocp.vars.E_BSPb = casadi.SX.sym('E_BSPb');
ocp.vars.E_BSLi = casadi.SX.sym('E_BSLi');
..... and some more symbolic variable declerations
..... (some more code)

ocp.sys.f = casadi.Function('f', {ocp.sys.x, ocp.sys.u}, {ocp.sys.rhs});
..... (some more code)
for k = 1:ocp.N
    ocp.lftsys.X(:, k+1) = ocp.sys.f(ocp.lftsys.X(:, k), ocp.lftsys.U(:, k));
end
.....
ocp.lp_prob = struct('f', ocp.J, 'x', ocp.lftsys.U, 'g', ocp.cnstr.g, 'p', ocp.lftsys.P);
ocp.solver = casadi.nlpsol('solver', 'ipopt', ocp.lp_prob, ocp.solver_opts);
.....
for k = 1:prms.sim.n_days * 24 / prms.sim.dt
    .....
    ocp.sol = ocp.solver('x0', ocp.sol_guess, 'lbx', ocp.cnstr.lbu, 'ubx', ocp.cnstr.ubu,...
                         'lbg', ocp.cnstr.lbg, 'ubg', ocp.cnstr.ubg, 'p', ocp.prmvals);
    .....
end
```

I thought of generating code for the casadi.Function(), and then somehow assigning that generated function to ocp.sys.f. But I dont even know if such an assignment is feasible.

Does someone know what would be the best way to generate C code from this code?

Thanks very much :)

**Comment by jgillis** [comment]
> it only generates C code from a function, is it correct?

Indeed. Note that NLP solvers are also CasADi Functions, so it is possible to generate the solver plus its dependencies.
ocp.solver.generate('solver.c')
As of CasADi 3.6, this only works for the `sqpmethod`, though.

If the goal is just to speed up certain parts, see https://github.com/casadi/casadi/wiki/FAQ:-how-to-perform-jit-for-function-evaluations-of-my-optimization-problem%3F

**Reply by Kamal-a-1** [reply]
Thanks very much for the reply.

Actually the goal is to run the generated C code on hardware in real-time.
I could use `solver.generate_dependencies('solver.c')` followed by `solver = nlpsol('solver', 'ipopt', './solver.dll')`, as in "nlp_codegen.py" example. When running the solver, I use 
```
ocp.sol = ocp.solver('x0', ocp.sol_guess, 'lbx', ocp.cnstr.lbu, 'ubx', ocp.cnstr.ubu,...
                          'lbg', ocp.cnstr.lbg, 'ubg', ocp.cnstr.ubg, 'p', ocp.prmvals);
```

By doing this I can avoid using casadi,Function() in MATLAB and that is very nice. But in order to create those constraints (like _ocp.constr.g_, etc.), I still need to use CasADi expressions. And when converting the rest of the code into C using MATLAB Coder, it finds a lot of unsupported statements even in `casadi.SX.sym()`. Do you know if there is any way to solve this?
